core/src/connection/pool.rs

```diff
@@ -0,0 +1,907 @@
+    ///
+    /// If a connection ID is given and no pending outgoing connection with that
+    /// ID is known to the pool, `None` is returned.
+    ///
+    /// If no connection ID is given and the pool contains at least one
+    /// pending outgoing connection to the given peer, an unspecified selection
+    /// is made among these connections.
+    ///
+    /// If no connection ID is given and no pending outgoing connection exists
+    /// to the given peer, `None` is returned.
+    pub fn get_outgoing(&mut self, peer: &TPeerId, id: Option<ConnectionId>)
+        -> Option<PendingConnection<'_, TInEvent, TConnInfo, TPeerId>>
+    {
+        match id {
+            Some(id) => match self.get(id) {
+                Some(PoolConnection::Pending(c)) => Some(c),
+                _ => None
+            }
+            None => self.pending.iter()
+                .find_map(|(id, (_endpoint, peer2))|
+                    if Some(peer) == peer2.as_ref() {
+                        Some(*id)
+                    } else {
+                        None
+                    })
+                .and_then(move |id| match self.manager.entry(id) {
+                    Some(manager::Entry::Pending(entry)) =>
+                        Some(PendingConnection {
+                            pending: &mut self.pending,
+                            entry,
+                        }),
+                    _ => None
+                })
+        }
+    }
+
+    /// Returns true if we are connected to the given peer.
+    ///
+    /// This will return true only after a `NodeReached` event has been produced by `poll()`.
+    pub fn is_connected(&self, id: &TPeerId) -> bool {
+        self.established.contains_key(id)
+    }
+
+    /// Returns the number of connected peers, i.e. those with at least one
+    /// established connection in the pool.
+    pub fn num_connected(&self) -> usize {
+        self.established.len()
+    }
+
+    /// Close all connections to the given peer.
+    pub fn disconnect(&mut self, peer: &TPeerId) {
+        if let Some(conns) = self.established.get(peer) {
+            for id in conns.keys() {
+                match self.manager.entry(*id) {
+                    Some(manager::Entry::Established(e)) => { e.close(); },
+                    _ => {}
+                }
+            }
+        }
+
+        for (id, (_endpoint, peer2)) in &self.pending {
+            if Some(peer) == peer2.as_ref() {
+                match self.manager.entry(*id) {
+                    Some(manager::Entry::Pending(e)) => { e.abort(); },
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    /// Asynchronously notifies a connection handler for the given peer of an event.
+    ///
+    /// If multiple connections exist to the peer, it is unspecified which handler
+    /// receives the event.
+    pub fn notify_handler(&mut self, peer: &TPeerId, event: TInEvent)
```

Comment:
Same remark. I think we will run into issues if we just leave selecting the connection to unspecified behaviours. This is the low-level API where we everything has to be strictly defined and we don't try to provide convenience methods.

Reply:
Done.